src/fwds_rvs_data.jl

```diff
@@ -861,12 +861,16 @@ tangent type. This method must be equivalent to `tangent_type(_typeof(primal))`.
 @foldable tangent_type(::Type{F}, ::Type{NoRData}) where {F<:Array} = F
 
 # Union types. NOTE: Only `Union{Nothing, Array{<:Base.IEEEFloat}, Base.IEEEFloat}` are supported. 
-@foldable tangent_type(F::Type{NoFData}, R::Type{Union{NoRData,Base.IEEEFloat}}) = Union{
-    NoTangent,tangent_type(Base.IEEEFloat)
-}
-@foldable tangent_type(F::Type{Union{NoFData,Array{<:Base.IEEEFloat}}}, R::Type{NoRData}) = Union{
-    NoTangent,tangent_type(Array{<:Base.IEEEFloat})
-}
+for T in [Float16, Float32, Float64]
+    @eval @foldable tangent_type(::Type{NoFData}, ::Type{Union{NoRData,$T}}) = Union{
+        NoTangent,tangent_type($T)
+    }
+    for N in 0:5  # Just go up to N=5 until general solution to https://github.com/chalk-lab/Mooncake.jl/pull/620 available
+        @eval @foldable tangent_type(::Type{Union{NoFData,Array{$T,$N}}}, ::Type{NoRData}) = Union{
+            NoTangent,tangent_type(Array{$T,$N})
+        }
+    end
+end
```

Comment:
Pretty sure it would be nicer to the compiler not to eval so many duplicate methods. Something like this, for example:

```suggestion
let f(@nospecialize x::Type) = Type{Union{NoRData, x}},
    f16 = f(Float16),
    f32 = f(Float32),
    f64 = f(Float64)
    global tangent_type
    @foldable function tangent_type(::Type{NoFData}, u::Union{f16, f32, f64})
        if u === f16
            t = Float16
        elseif u === f32
            t = Float32
        elseif u === f64
            t = Float64
        end
        Union{NoTangent, tangent_type(t)}
    end
end

for T in [Float16, Float32, Float64]
    for N in 0:5  # Just go up to N=5 until general solution to https://github.com/chalk-lab/Mooncake.jl/pull/620 available
        @eval @foldable tangent_type(::Type{Union{NoFData,Array{$T,$N}}}, ::Type{NoRData}) = Union{
            NoTangent,tangent_type(Array{$T,$N})
        }
    end
end
```